File: ChMLab/simpleiterations.js
var a = [
    [-0.95, -0.06, -0.12, 0.14],
    [0.04, -1.12, 0.68, 0.11],
    [0.34, 0.08, -1.06, 0.44],
    [0.11, 0.12, 0, -1.03]
];
var r = [2.17, -1.4, 2.1, 0.8];

function base(x) {
    return [
        (r[0] - a[0][1] * x[1] - a[0][2] * x[2] - a[0][3] * x[3]) / a[0][0],
        (r[1] - a[1][0] * x[0] - a[1][2] * x[2] - a[1][3] * x[3]) / a[1][1],
        (r[2] - a[2][0] * x[0] - a[2][1] * x[1] - a[2][3] * x[3]) / a[2][2],
        (r[3] - a[3][0] * x[0] - a[3][1] * x[1] - a[3][2] * x[2]) / a[3][3]
    ];
}

function calc(ebs) {
    var res=[0, 0, 0, 0];
    var temp;
    do
    {
        temp=arraySum(res);
        res = base(res);
    }while (Math.abs(arraySum(res)-temp)>(1-form1())/form1()*ebs);
    return res;
}

function arraySum(a) {

    var sum = 0;
    for (var i=0; i<a.length; i++) {
        sum += a[i];
    }
    return sum;

}

function form1() {
    var sum = 0, sumt = 0;
    for (var i = 0; i<4; i++)
    {
        sumt = 0;
        for (var j=0; j<4; j++)
        {
            if (i != j)
                sumt += Math.abs(a[i][j] / a[i][i]);
        }
        if (sum <= sumt) {
            sum = sumt;
        }
    }
    return sum;
}

var p=calc(0.001);
console.log(p[0] + "  "+  p[1] + "  "+  p[2] + "  " + p[3]);
//console.log(p);

File: ChM/lab1/lab1/Algorithms/SimpleIterations.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;

namespace lab1.Algorithms
{
    public class SimpleIterations
    {
        public double[][] A { get; set; }
        public double[] B { get; set; }

        public double[][] _B { get; set; }
        public double[] _D { get; set; }

        public List<string> Log { get; set; }

        public SimpleIterations(int n)
        {
            A = new double[n][];
            _B = new double[n][];
            for (int i = 0; i < n; i++)
            {
                A[i] = new double[n];
                _B[i] = new double[n];
            }
            B = new double[n];
            _D = new double[n];
            Log = new List<string>();
        }

        private double[] Base(double[] x)
        {
            int n = x.Length;
            double[] newX = new double[n];

            for (int i = 0; i < n; i++)
                newX[i] = 0;

            for (int i = 0; i < n; i++)
            {
                newX[i] += B[i];
                for (int j = 0; j < n; j++)
                {
                    if (j != i)
                        newX[i] -= A[i][j] * x[j];
                }
                newX[i] /= A[i][i];
            }

            return newX;
        }

        private double FirstForm()
        {
            double sum = 0, sumt = 0;
            int n = A.Length;
            for (int i = 0; i < n; i++)
            {
                sumt = 0;
                for (int j = 0; j < n; j++)
                {
                    if (i != j)
                        sumt += Math.Abs(A[i][j] / A[i][i]);
                }
                if (sum < sumt)
                    sum = sumt;
            }
            return sum;
        }


        public double[] Calculate(double eps)
        {
            Log.Clear();
            int n = A.Length;
            for (int i = 0; i < n; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    if (i != j)
                        _B[i][j] = A[i][j] / A[i][i];
                    else
                        _B[i][j] = 0;
                }
                _D[i] = B[i] / A[i][i];
            }

            double b1 = FirstForm();

            if (b1 >= 1)
            {
                Log.Add("Данная система не сходится по этому алгоритму");
                return null;
            }

            double[] res = _D;
            double[] rest;
            int k = 1;
            string s = "Итерация №0\n";
            s += string.Format("x1 = {0}; x2 = {1}; x3 = {2}; x4 = {3}\n********\n",
                Math.Round(res[0], (eps % 1).ToString().Length - 2), Math.Round(res[1], (eps % 1).ToString().Length - 2),
                Math.Round(res[2], (eps % 1).ToString().Length-2), Math.Round(res[3], (eps % 1).ToString().Length - 2));
            Log.Add(s);
            do
            {
                s = "";
                rest = res;
                res = Base(res);
                s += "Итерация №" + k + "\n";
                s += string.Format("x1 = {0}; x2 = {1}; x3 = {2}; x4 = {3}\n********\n",
                    Math.Round(res[0], (eps % 1).ToString().Length - 2), Math.Round(res[1], (eps % 1).ToString().Length - 2), 
                    Math.Round(res[2], (eps % 1).ToString().Length - 2), Math.Round(res[3], (eps % 1).ToString().Length - 2));
                Log.Add(s);
                k++;
            } while (Math.Abs(res.Sum() - rest.Sum()) > Math.Abs(1 - b1) / b1 * eps);

            
            return res;
        }
    }

}
